### PR TITLE
Tracer wrapping

### PR DESCRIPTION
**Pull request by jcchavezs**
This PR attempts to start using the tracer in the test. Currently it only supports options from the zipkin native tracer but some options we could include in the future (if there is user requests for it) are:
- span observers
- dropAllLogs
- MaxLogSperSpan

**One big issue we found was the fact that OT defines the sampling at tag level so we might need to mutate the context but we don't have access to it. We need to iterate over it.**

Ping @basvanbeek @stakhiv

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,15 +1,17 @@
 language: go
 
 go:
-  - 1.8.x
-  - 1.9.x
-  - 1.10.x
+  # - 1.8.x
+  # - 1.9.x
+  # - 1.10.x
```
Comment (jcchavezs) [inline .travis.yml]: We disabled this until https://github.com/openzipkin/zipkin-go/pull/113 gets merged.

File: bench_test.go [hunk 2]
```diff
@@ -79,25 +94,24 @@ func BenchmarkSpan_100BaggageItems(b *testing.B) {
...
-	}, 100, 100, 100)
-	if int(r) != b.N {
-		b.Fatalf("missing traces: expected %d, got %d", b.N, r)
-	}
-}
+// func BenchmarkTrimmedSpan_100Events_100Tags_100BaggageItems(b *testing.B) {
+// 	var r CountingRecorder
+// 	t, err := NewTracer(
+// 		&r,
+// 		zipkin.WithSampler(zipkin.NeverSample),
```
Comment (jcchavezs) [inline bench_test.go]: As far as I see on current zipkin-go-ot, even when noop the span gets into the recorder, is that true? https://github.com/openzipkin-contrib/zipkin-go-opentracing/blob/master/span_test.go#L167

**Comment by basvanbeek** [comment]
> One big issue we found was the fact that OT defines the sampling at tag level so we might need to mutate the context but we don't have access to it. We need to iterate over it.

If you are talking about the `sampling.priority` tag this is "optional"

> If greater than 0, a hint to the Tracer to do its best to capture the trace. If 0, a hint to the trace to not-capture the trace. If absent, the Tracer should use its default sampling mechanism.

I'd be perfectly OK with not implementing this for the first iteration (we don't support it in the native implementation at the moment either) and nothing wrong with... "we did our best but no" :)

If we want to support you'd typically not iterate over the tag keys but on key insertion would take action by altering the sample value in context (obviously this wouldn't work if allowing NoopSpan)

**Comment by jcchavezs** [comment]
Right @basvanbeek so the issue here is that even if the tag is optional, conceptually zipkin has the decision for sampling on beforehand whereas a tag can be set at any moment in the lifecycle of the span. I honestly prefer the approach of not supporting this and maybe add a comment in the code saying that we don't support it. Also as you say, if someone decides to not spawn a real span there is no way we can switch to a real span from a nop span. Definitively something we should add to documentation. Ping @stakhiv

**Comment by jcchavezs** [comment]
Since https://github.com/openzipkin/zipkin-go/pull/113 is merged we can reenable old versions and fix the gomod to master. After that we can merge it @stakhiv

**Comment by basvanbeek** [comment]
> Right @basvanbeek so the issue here is that even if the tag is optional, conceptually zipkin has the decision for sampling on beforehand whereas a tag can be set at any moment in the lifecycle of the span. I honestly prefer the approach of not supporting this and maybe add a comment in the code saying that we don't support it. Also as you say, if someone decides to not spawn a real span there is no way we can switch to a real span from a nop span. Definitively something we should add to documentation. Ping @stakhiv

Hence it being a hint because previously spawned client calls would have the same outdated sampling info propagated once the tag is set at a later stage. I'm not saying implement, I'm saying if we would implement our guarantees would not be much different from other OT implementations (unless it involves a post sampling ecosystem)

**Comment by jcchavezs** [comment]
So after a chat with Bas, we agreed that we need our own functional options
for this tracer so we provide a consistent experience for developers.

José Carlos Chávez


ons. 15. mai 2019 kl. 13:02 skrev Bas van Beek <notifications@github.com>:

**Pull request merged**